pandas/core/arrays/datetimelike.py

```diff
@@ -2217,11 +2214,11 @@ def validate_periods(periods: None) -> None:
 
 
 @overload
-def validate_periods(periods: float) -> int:
+def validate_periods(periods: int | float) -> int:
     ...
 
 
-def validate_periods(periods: float | None) -> int | None:
+def validate_periods(periods: int | float | None) -> int | None:
```

Comment:
@twoertwein so the typing simplification(?) that converted `int | float` -> `float` isn't an enforced typing check?

Reply:
flake8-pyi does that check but only for pyi files not py files.

More generally, we have many Unions that contain redundant classes, for example Scalar contains datetime and pd.Timestamp but I think pd.Timestamp inherits from datetime. That is not a problem (also type checkers might need some more time) but it is not good style.

Reply:
Should I revert this?  i find it clearer, but not a big deal.

Reply:
I wouldn't mind ignoring [flake8-pyi](https://github.com/PyCQA/flake8-pyi)'s Y041 when we run flake8-pyi: then py and pyi files are more consistent and there is also some push back against it https://github.com/python/mypy/issues/13760#issuecomment-1262523887

Reply:
btw in a world where `float` is equivalent to `int | float`, how would i express "specifically a python float object and not an integer object"

Reply:
In the type checkers world, you can't: https://peps.python.org/pep-0484/#the-numeric-tower

> Rather than requiring that users write import numbers and then use numbers.Float etc., this PEP proposes a straightforward shortcut that is almost as effective: when an argument is annotated as having type float, an argument of type int is acceptable; similar, for an argument annotated as having type complex, arguments of type float or int are acceptable. This does not handle classes implementing the corresponding ABCs or the fractions.Fraction class, but we believe those use cases are exceedingly rare.

But I wonder how python and its type system would have evolved if they started together - we hopefully wouldn't have this mess of type checkers/PEPs dictating that int is treated as a subclass of float but the implementation doesn't do that.

Reply:
that seems weird given that int _isnt_ a subclass of float.  anyway, shouldnt be relevant for this PR